sdk/src/types/block/address/restricted.rs

```diff
@@ -0,0 +1,294 @@
+#[repr(transparent)]
+pub struct AddressCapabilities(BoxedSlicePrefix<u8, u8>);
+
+impl AddressCapabilities {
+    pub fn all() -> Self {
+        let mut res = Self::default();
+        res.set_all();
+        res
+    }
+
+    pub fn none() -> Self {
+        Self::default()
+    }
+
+    pub fn is_all(&self) -> bool {
+        AddressCapabilityFlag::all().all(|flag| self.has_capability(flag))
+    }
+
+    pub fn is_none(&self) -> bool {
+        self.0.iter().all(|b| 0.eq(b))
+    }
+
+    pub fn set_all(&mut self) -> &mut Self {
+        for flag in AddressCapabilityFlag::all() {
+            self.add_capability(flag);
+        }
+        self
+    }
+
+    pub fn set_none(&mut self) -> &mut Self {
+        *self = Default::default();
+        self
+    }
+
+    pub fn add_capability(&mut self, flag: AddressCapabilityFlag) -> &mut Self {
+        if self.0.len() <= flag.index() {
+            let mut v = Box::<[_]>::from(self.0.clone()).into_vec();
+            v.resize(flag.index() + 1, 0);
+            // Unwrap: safe because the indexes are within u8 bounds
+            self.0 = v.into_boxed_slice().try_into().unwrap();
+        }
+        self.0[flag.index()] |= flag.as_byte();
+        self
+    }
+
+    pub fn add_capabilities(&mut self, flags: impl IntoIterator<Item = AddressCapabilityFlag>) -> &mut Self {
+        for flag in flags {
+            self.add_capability(flag);
+        }
+        self
+    }
+
+    pub fn with_capabilities(mut self, flags: impl IntoIterator<Item = AddressCapabilityFlag>) -> Self {
+        self.add_capabilities(flags);
+        self
+    }
+
+    pub fn set_capabilities(&mut self, flags: impl IntoIterator<Item = AddressCapabilityFlag>) -> &mut Self {
+        *self = Self::default().with_capabilities(flags);
+        self
+    }
+
+    pub fn has_capability(&self, flag: AddressCapabilityFlag) -> bool {
+        self.0
+            .get(flag.index())
+            .map(|byte| byte & flag.as_byte() == flag.as_byte())
+            .unwrap_or_default()
+    }
+
+    pub fn has_capabilities(&self, flags: impl IntoIterator<Item = AddressCapabilityFlag>) -> bool {
+        flags.into_iter().all(|flag| self.has_capability(flag))
+    }
+
+    pub fn split(&self) -> impl Iterator<Item = AddressCapabilityFlag> + '_ {
+        self.0.iter().enumerate().flat_map(|(idx, byte)| {
+            AddressCapabilityFlag::all().filter(move |f| (idx == f.index() && byte & f.as_byte() == f.as_byte()))
+        })
+    }
+}
+
+impl<I: IntoIterator<Item = AddressCapabilityFlag>> From<I> for AddressCapabilities {
+    fn from(value: I) -> Self {
+        Self::default().with_capabilities(value)
+    }
+}
+
+impl Packable for AddressCapabilities {
+    type UnpackError = Error;
+    type UnpackVisitor = ();
+
+    fn pack<P: packable::packer::Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
+        if !self.is_none() {
+            self.0.pack(packer)?;
+        } else {
+            0_u8.pack(packer)?;
+        }
```

Comment:
Can this all not simply be `self.0.pack(packer)?;` ?

Reply:
the issue is that we aren't supposed to serialize the flags if they are all false